File: scrapers/bobaedream.py
import httpx

from scrapers.base import BaseScraper, Post


class BobaedreamScraper(BaseScraper):
    community = "bobaedream"
    community_name = "보배드림"
    base_url = "https://www.bobaedream.co.kr"

    async def fetch_best_posts(self, client: httpx.AsyncClient) -> list[Post]:
        soup = await self._fetch_html(client, f"{self.base_url}/list?code=best")
        posts = []

        table = soup.find("table", class_="clistTable02")
        if not table:
            return []

        for tr in table.find_all("tr"):
            a_tag = tr.select_one("td.pl14 a")
            if not a_tag:
                continue
            title = a_tag.get_text(strip=True)
            href = a_tag.get("href", "")
            if not href or not title:
                continue

            votes = 0
            rec_td = tr.select_one("td.recomm")
            if rec_td:
                try:
                    votes = int(rec_td.get_text(strip=True).replace(",", ""))
                except ValueError:
                    pass

            views = 0
            cnt_td = tr.select_one("td.count")
            if cnt_td:
                try:
                    views = int(cnt_td.get_text(strip=True).replace(",", ""))
                except ValueError:
                    pass

            comments = 0
            reply_span = tr.select_one("span.totreply")
            if reply_span:
                try:
                    comments = int(reply_span.get_text(strip=True))
                except ValueError:
                    pass

            url = href if href.startswith("http") else f"{self.base_url}{href}"
            posts.append(self._make_post(title, url, votes=votes, views=views,
                                         comments=comments))

        return posts

File: internal/scraper/bobaedream.go
package scraper

import (
	"fmt"
	"net/http"
	"strconv"
	"strings"

	"github.com/PuerkitoBio/goquery"
)

// BobaedreamScraper scrapes 보배드림 best posts.
type BobaedreamScraper struct {
	baseScraper
}

func NewBobaedreamScraper() *BobaedreamScraper {
	return &BobaedreamScraper{
		baseScraper: baseScraper{
			community:     "bobaedream",
			communityName: "보배드림",
			baseURL:       "https://www.bobaedream.co.kr",
			encoding:      "utf-8",
		},
	}
}

func (s *BobaedreamScraper) Name() string { return s.communityName }

func (s *BobaedreamScraper) FetchBestPosts(client *http.Client) ([]Post, error) {
	doc, err := fetchDocument(client, fmt.Sprintf("%s/list?code=best", s.baseURL), s.encoding, s.baseURL)
	if err != nil {
		return nil, err
	}

	var posts []Post

	table := doc.Find("table.clistTable02").First()
	if table.Length() == 0 {
		return posts, nil
	}

	table.Find("tr").Each(func(_ int, tr *goquery.Selection) {
		aTag := tr.Find("td.pl14 a").First()
		if aTag.Length() == 0 {
			return
		}
		title := strings.TrimSpace(aTag.Text())
		href, exists := aTag.Attr("href")
		if !exists || href == "" || title == "" {
			return
		}

		votes := 0
		recTd := tr.Find("td.recomm").First()
		if recTd.Length() > 0 {
			text := strings.TrimSpace(strings.ReplaceAll(recTd.Text(), ",", ""))
			if v, err := strconv.Atoi(text); err == nil {
				votes = v
			}
		}

		views := 0
		cntTd := tr.Find("td.count").First()
		if cntTd.Length() > 0 {
			text := strings.TrimSpace(strings.ReplaceAll(cntTd.Text(), ",", ""))
			if v, err := strconv.Atoi(text); err == nil {
				views = v
			}
		}

		comments := 0
		replySpan := tr.Find("span.totreply").First()
		if replySpan.Length() > 0 {
			if v, err := strconv.Atoi(strings.TrimSpace(replySpan.Text())); err == nil {
				comments = v
			}
		}

		url := href
		if !strings.HasPrefix(href, "http") {
			url = s.baseURL + href
		}
		posts = append(posts, s.makePost(title, url, votes, views, comments))
	})

	return s.filterPosts(posts), nil
}
